### ISSUE TITLE
IX: show training data dashboard

### ISSUE THREAD

**Issue by txau** [huridocs/uwazidocs#4432]
Display atop of the table, before the column headers a mini-dashboard with:
- Amount of training samples (label match + label mismatch + label empty)
- Amount of suggestion slots (number of documents in each language)
- Amount of non-obsolete suggestions

Keep it updated in real time as the users label data or the model sends suggestions.

**Comment by gabriel-piles** [comment]
@txau Can we add to the dashboard the time and date when find suggestions was pressed or the machine learning model was trained?

**Comment by txau** [comment]
@gabriel-piles good point.

**Comment by juanmnl** [comment]
@txau can I get some real values for those amounts (ie. 10 - 10.000?) tx!

**Comment by txau** [comment]
@juanmnl Based on real data. The total amount is 8900 (slots). Out of these I have around 50 labeled samples that are used for training and 300 that match but have no text selection so they cannot be used as training data.

After last training was complete:
- Training samples 50: they have a value and text selection. This tells the user how much training data will be used.
- Matching 300: they have a matching value and they may have or not a text selection. This let's the user know how good the model is at predicting the right value based on the current training set. There may be overlapping with the training samples.
- Not matching 1000: they have a current value and a predicted value but they do not match. This also indicates the user how good the model is at predicting the right value. There may be overlapping with the training samples.
- Empty 7600: they have no current value or an empty model suggestion or both are empty. They can overlap with the training samples but do not overlap with the matching/not matching.

So, something like this:

![image](https://user-images.githubusercontent.com/489831/159458201-c6d86aaa-3754-4543-8928-cab973385aea.png)

Which is telling the user: You need more training data, most of the model predictions do not match and the model is not being able to pull many non-empty suggestions.

Imagine the user labels some more data and it starts to look better:

![image](https://user-images.githubusercontent.com/489831/159458780-a2b7eb94-526f-46c0-a8ef-b0c2933d92e3.png)

Showing that without a lot of training data the model is able to correctly predict most of the labels, but there are still a considerable quantity of non-matching and empty suggestions. If the user labels a little bit more training data (specially focusing in the empty and non-matching cases) and re-trains:

![image](https://user-images.githubusercontent.com/489831/159459388-75bb623e-0072-4af5-b79b-fac9c3db2e98.png)

This is a healthy state. With relatively low training samples the model is able to correctly predict most values, leaving some outsiders and edge cases out (to be manually filled or add more training data).

**Comment by txau** [comment]
https://www.figma.com/file/VbnNs1Iy8qbOet3oLwMgfe/Information-Extraction?node-id=250%3A126

**Comment by txau** [comment]
@fnocetti picking up the color code in this frame instead of the previous one: https://www.figma.com/file/VbnNs1Iy8qbOet3oLwMgfe/Information-Extraction?node-id=256%3A1206

The desired behavior is:
- RED: represents "match/label" state alone. That is, the training samples and whether it is actually being able to train the model and return the same suggestion as it was trained for.
- BLUE: represents all the labeled data, that is "match/label" + "mismatch/label" + "empty/label". In the graphic this means the red squares need to overlap the blue ones with respect to the total amount of suggestions, since blue is always going to be equal or greater than red (blue contains red + other cases), but they are both training data.

The combination of red and blue act as a self-test of the model. In an ideal scenario, there is no blue and just red, meaning that all training data is also processed properly giving the expected results for the model. A big difference between blue and red can be caused by erratic training data with unbalanced datasets among other causes.

In fact I would suggest that we invert the color code for blue being "match/label" and red representing the mismatches in labeled data since red tends to be a warning color.

The rest of the data does not overlap with the red/blue area and needs to be divided in:
- GREEN: matching suggestions, that is, non-labeled data that is matching the suggested value. This is good and it is what we are looking for. The more green the better. It represents the state "match/value".
- YELLOW: mismatching suggestions not including the state "mismatch/label", so it represents the states "mismatch/value" and "mismatch/empty". Yellow is bad, the less yellow the better.
- GREY: everything else. Grey is ok if the data is being processed (obsolete) or if we have low training samples. Grey is bad if we have a lot of training sample but it is not able to figure out the right suggestions.

Green, yellow and grey do not overlap.

As for the legend, it should be:
- Blue for "Training". All the training samples including mismatched and empty ones.
- Green for "Matching". Including labeled data, that is "match/label" + "match/value".
- Yellow for "Non-matching". Including labeled data, that is "mismatch/label" + "mismatch/value" + "mismatch/empty" + "empty/value" and "empty/label".
- Grey for "Empty/Obsolete", that is "empty/empty" and "obsolete".

**Comment by fnocetti** [comment]
Keeping this open as we test and review the semantics of the chart.

**Comment by txau** [comment]
Feedback on the current implementation:
- Have the dash-board self-update without needing to reload. This can be achieved with a timer or with the events from training.
- Simplify the training health dashboard:
  - Remove red squares. Blue now should represent all labeled data (label/match + label/mismatch + label/empty)
  - Green should represent "match/value" only
  - Yellow should represent "mismatch/value" only
  - Grey should include "mismatch/empty" (previously included in yellow).
- Introduce a new label, in grey "Accuracy:" with the value "match/value divided by match/value + mismatch/value"

Also, I think there is something off with the initial state. After the model is trained and all suggestions are configured as "obsolete", they do not represent the 100% of the suggestions.
![image](https://user-images.githubusercontent.com/489831/179795378-fa9fd4a1-7ce3-4d90-b9c1-a23e0e39e8e7.png)

**Comment by txau** [comment]
@fnocetti another thing that I noticed is the blank state showing 0 for the training data. I think it should show the training amount since it is unrelated to matching or obsolescence. The data in the dashboard can change as the find suggestion process advances, but also when the user is interacting with the data (adding labeled data, adding values, etc). So maybe the dashboard should get updated with a timer or also triggered by other events).

I was also wondering if the miscount in the total amount of obsolete (the image in the previous post) is related to the problem with the documents not having a language as explained in #4876 .

**Comment by fnocetti** [comment]
@txau 
> @fnocetti another thing that I noticed is the blank state showing 0 for the training data. I think it should show the training amount since it is unrelated to matching or obsolescence. 

It's not unrelated. The counts are based on the state of the suggestions, and things that were matching before, are changed to state Obsolete when clicking on the button, so cannot be counted any longer. 
Anyway, if something can be Obsolete but still Matching, and Labeled, to me that's definitively a smell that the state should actually be at least 2 different fields (matching state and labeling state, or whatever names), and also that some states like Obsolete and Error should not be a value for any of those fields, but possibly for a third one (maybe what we call status right now). Sums up to 3 fields.

> The data in the dashboard can change as the find suggestion process advances, but also when the user is interacting with the data (adding labeled data, adding values, etc). So maybe the dashboard should get updated with a timer or also triggered by other events).

- It currently changes as the user interacts with the application when accepting suggestions, or when labeling data through the PDF view in the suggestions list. If it doesn't work, that's a bug.
- Updating as suggestions arrive after clicking the find suggestions button is not implemented and should be rather quick to do.
- On the other side, updating because of actions that happen in other areas of the application (like setting values through entity editing) will need websocket events to be properly done. Now we have the event bus, we can achieve this with reasonable efforts by emitting a IX specific message through the WS. Something like `hey, there's new IX info, you should fetch it`. Using timers is a quicker workaround, but hacky approach and can unnecessarily use server's and mongo's cpu.

> I was also wondering if the miscount in the total amount of obsolete (the image in the previous post) is related to the problem with the documents not having a language as explained in #4876 .

There are several things to consider there:
- The `Error` state is not considered in any count. So it may be that the remaining 8,26% of the suggestions are in `Error` state.
- It may be the case the they are in the `Error` **status** (different than state). The query fetching this numbers is fetching every suggestion for the property, without caring about the status. I'm wondering if it can happen that those suggestions are in status Error while keeping the state Obsolete just as garbage data? If that's the case, we may want to check how that happened, but also should take a look at if we should care for the status in this query (the list of suggestions does care, for example)

**Comment by txau** [comment]
1. Got it.

2. I don't think we need to account for the events from other parts of the application at this stage. If the dashboard is being updated by interactions with data in the IX module + finding suggestions process is more than enough.

**Comment by txau** [comment]
> There are several things to consider there:
> 
>     * The `Error` state is not considered in any count. So it may be that the remaining 8,26% of the suggestions are in `Error` state.
> 
>     * It may be the case the they are in the `Error` **status** (different than state). The query fetching this numbers is fetching every suggestion for the property, without caring about the status. I'm wondering if it can happen that those suggestions are in status Error while keeping the state Obsolete just as garbage data? If that's the case, we may want to check how that happened, but also should take a look at if we should care for the status in this query (the list of suggestions does care, for example)

I didn't find any suggestions with error as state or status. Another more likely reason is that screenshot is taken while the first batch is being processed, so these first 50 suggestions are in a state that is not going to be added to any stat in the legend. In fact this collection has 605 files and the count is showing 555 (605 - 50 from the current batch). 

If you see a quick fix so we never lose sight of the real number (605) go for it. If it is too much effort I think we can live with it since at the end of the process we get the right totals.
